lib/link.js

```diff
@@ -100,11 +107,13 @@ Link.prototype.attach = function() {
 
 Link.prototype.detach = function() {
   var self = this;
+  this._timeouts = undefined; // Disable any re-attachment policy.
   var detachPromise = new Promise(function(resolve, reject) {
     var onError = function(err) { reject(err); };
     self.once(Link.ErrorReceived, onError);
     self.once(Link.Detached, function(info) {
       self.removeListener(Link.ErrorReceived, onError);
+      self.session._removeLink(self);
```

Comment:
something feels wrong about this, shouldn't the session listen to Link.Detach events itself and remove the links from there? I guess this is better stated as: why is Session's internal state something Link cares about?